### ISSUE TITLE
TransformerDecoder produces identical outputs regardless of input

### ISSUE THREAD

**Issue by nicolacalzone** [pytorch/pytorch#149886]
### 🐛 Describe the bug
I'm building a decoder-only transformer using TransformerDecoder and TransformerDecoderLayer classes from the PyTorch library. 
The model consistently produces the same output tensor regardless of the input, also with all rows in the output being identical.

### My environment:
```
Python 3.10.12
torch 2.5.1
Ubuntu 22.04
```

### Minimal Reproduction Code

```python
class DecoderOnlyTransformer(pl.LightningModule):
    def __init__(
        self, 
        vocab_size, 
        max_seq_length, 
        d_model=512, 
        nhead=8, 
        num_decoder_layers=6, 
        dim_feedforward=2048, 
        dropout=0.1,
        learning_rate=5e-4,
        optimizer_type="adam"
    ):
        super().__init__()
        self.save_hyperparameters()
        self.vocab_size = vocab_size
        self.d_model = d_model
        self.learning_rate = learning_rate
        self.optimizer_type = optimizer_type
        
        # Model components
        self.embeddings = nn.Embedding(vocab_size, d_model)
        self.pos_embeddings = PositionalEmbedding(d_model, max_seq_length)

        self.single_decoder_layer = nn.TransformerDecoderLayer(d_model=d_model,
                                                         nhead=nhead,
                                                         dropout=dropout,
                                                         dim_feedforward=dim_feedforward,
                                                         activation="gelu")
        
        self.stack_decoder_layers = nn.TransformerDecoder(self.single_decoder_layer, num_decoder_layers)
        self.output_projection = nn.Linear(d_model, vocab_size)

    def forward(self, tgt, tgt_mask=None, memory_mask=None,
                tgt_key_padding_mask=None, memory_key_padding_mask=None) -> torch.Tensor:
        
        tgt = self.embeddings(tgt)
        tgt = self.pos_embeddings(tgt)

        
        # Generate causal mask if not provided
        if tgt_mask is None:
            seq_len = tgt.size(0)
            tgt_mask = nn.Transformer.generate_square_subsequent_mask(seq_len).to(tgt.device)

        memory = torch.zeros(1, tgt.size(1), self.embeddings.embedding_dim).to(tgt.device)
            
        output = self.stack_decoder_layers(
            tgt,
            memory,
            tgt_mask=tgt_mask,
            memory_mask=memory_mask,
            tgt_key_padding_mask=tgt_key_padding_mask,
            memory_key_padding_mask=memory_key_padding_mask
        )
        return output

```

For any input, the model produces an output tensor where:

- All positions in the sequence have identical values
- The output is always the same regardless of input

Output:

```python
tensor([[[ 0.3030,  0.2662,  0.7048,  ...,  0.7949, -0.4710, -0.2264]],

        [[ 0.3030,  0.2662,  0.7048,  ...,  0.7949, -0.4710, -0.2264]],

        [[ 0.3030,  0.2662,  0.7048,  ...,  0.7949, -0.4710, -0.2264]],

        ...,

        [[ 0.3030,  0.2662,  0.7048,  ...,  0.7949, -0.4710, -0.2264]],

        [[ 0.3030,  0.2662,  0.7048,  ...,  0.7949, -0.4710, -0.2264]],

        [[ 0.3030,  0.2662,  0.7048,  ...,  0.7949, -0.4710, -0.2264]]],
       device='cuda:0')
```

### Expected Behavior
The decoder should produce different outputs for different inputs, with variations across sequence positions.

### Additional Context

- I'm following the PyTorch documentation for the TransformerDecoder implementation
- The issue persists regardless of input sequence or mask configuration
- The PositionalEmbedding implementation is standard (sinusoidal)

### Questions
- Is there something wrong with how I'm initializing or using the TransformerDecoder?
- Could the zero-initialized memory be causing this behavior? (I set it like that because there is no encoder, and from the documentation I happen to understand that memory is its output)
- Are there known issues with the TransformerDecoder implementation?

Any help diagnosing this would be greatly appreciated!

### Versions

Python 3.10.12

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by nicolacalzone** [comment]
It turns out that `TransformerDecoder` and `TransformerDecoderLayer` are not actually designed for decoder-only Transformer architectures, despite their naming.  

**Solution:**  
For a decoder-only model, you should instead use `TransformerEncoder` and `TransformerEncoderLayer` with the `causal` flag set to `True`.  

**Additional Note on Attention Masks:**  
I initially tried using a triangular attention mask (lower diagonal filled with `False`, upper diagonal with `True`), but it produced the same results as before with TransformerDecoder module (so: always same result at inference time). 
After removing the attention mask entirely, the Transformer began training correctly and learning as expected.  And no, you cannot remove the mask from the TransformerDecoderLayer. It is compulsory to use it there and using it filled with 0s or 1s does not work.
